### PR TITLE
Fixes for warnings generated by ma 4.0 and others.

### LINKED ISSUE #361: RemovedInMarshmallow4Warning: Passing field metadata as a keyword arg is deprecated. Use the explicit `metadata=...` argument instead.
### Env
marshmallow-sqlalchemy==0.24.1
marshmallow==3.10.0
sqlalchemy==1.3.22

### Issue

`ModelConverter._add_column_kwargs()` passes a `"places"` keyword argument to `fields.Float` which isn't part of the fields signature. Until now it would have been silently swallowed into field metadata, however it now generates a warning:

```
RemovedInMarshmallow4Warning: Passing field metadata as a keyword arg is deprecated. Use the explicit `metadata=...` argument instead.
```
It checks for the column type to have a `scale` property:

https://github.com/marshmallow-code/marshmallow-sqlalchemy/blob/ca4c3bec00db09b161ab1603ea2c23017753a758/src/marshmallow_sqlalchemy/convert.py#L277-L278

Both `Float` and `Numeric`/`Decimal` column types have the `scale` property, yet only the `Decimal` field type accepts the `places` arg.

The sqlalchemy [`Numeric`](https://github.com/sqlalchemy/sqlalchemy/blob/640cd8a70f8a664b7834c5f74ec322fdea644043/lib/sqlalchemy/sql/sqltypes.py#L587) type has the `asdecimal` flag:

>         :param asdecimal: default True.  Return whether or not
>           values should be sent as Python Decimal objects, or
>           as floats.   Different DBAPIs send one or the other based on
>           datatypes - the Numeric type will ensure that return values
>           are one or the other across DBAPIs consistently.

Changing the above check to this seems to work well:
```py
        if getattr(column.type, "asdecimal", False):
            kwargs["places"] = getattr(column.type, "scale", None)
```
...which only passes `places` to the field constructor when sqlalchemy is explicit about using a decimal type.

### PR DESCRIPTION
**Pull request by peterschutt**
- only pass `places` param to field when sure we are using a `Decimal`
field.
- limit where `info={"marshmallow: ...}"` is passed to `Column` in tests
in order to reduce incidence of depreciation alert when running tests.
- Use `metadata` keyword where passing non-field keywords to field
constructor.
- Remove test for `sa.Binary` as that column type is depreciated on the
sqlalchemy side.

### CODE REVIEW

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:55]: In order to bucket non-field arguments in to a `metadata` argument, we need to be able to differentiate which kwargs to `property2field()` and `column2field()` are field parameters and which aren't.

This creates a set of the names of each of the potential keyword parameters of the field class and its parents and then checks if each of the kwargs is a part of that set, or not, and allocates to `field_kwargs` or `field_kwargs["metadata"]` as appropriate.

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:176]: I wasn't totally sure how I should handle this, but doing it this way will put the `column` attribute into the `metadata` when `ret` is an instance of `Related` (as @AbdealiJK was experiencing in #361).

File: tests/conftest.py
Comment (peterschutt) [inline tests/conftest.py:57]: This usage is depreciated, so I've removed from here and I add a specific column to the model for the test of this specific behavior..

File: tests/test_conversion.py
Comment (peterschutt) [inline tests/test_conversion.py:148]: `sa.Binary` depreciated by `sqlalchemy`..

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:56]: In order to bucket non-field arguments in to a metadata argument, we need to be able to differentiate which kwargs to property2field() and column2field() are field parameters and which aren't.

This creates a set of the names of each of the potential keyword parameters of the field class and its parents and then checks if each of the kwargs is a part of that set, or not, and allocates to field_kwargs or field_kwargs["metadata"] as appropriate.

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:56]: In order to bucket non-field arguments in to a metadata argument, we need to be able to differentiate which kwargs to `property2field()` and `column2field()` are field parameters and which aren't.

This creates a set of the names of each of the potential keyword parameters of the field class and its parents, checks if each of the kwargs is a part of that set, or not, and allocates to `field_kwargs` or `field_kwargs["metadata"]` as appropriate.

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py]: Need to behave differently if `fields.Field(..., metadata=None, ...)` or `fields.Field(..., **metadata)`.

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:67]: Determine the field's possible keyword arguments and explicitly set arguments that aren't a part of that set as metadata.

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [inline src/marshmallow_sqlalchemy/convert.py:194]: Resolves issue that @AbdealiJK had in #361 where `column='name'` was causing depreciation warning when passed to `RelatedList`.

File: tests/conftest.py
Comment (peterschutt) [inline tests/conftest.py:57]: Passing `info=dict(marshmallow=...)` is depreciated so I remove from here and I add a column to test the behavior in the specific test. This stops warnings being generated for other tests that aren't testing that particular behavior.

File: tests/test_conversion.py
Comment (peterschutt) [inline tests/test_conversion.py:148]: This column type is depreciated by sqlalchemy.

File: tests/test_model_schema.py
Comment (peterschutt) [inline tests/test_model_schema.py:734]: Including the `"id"` keys here silences a sqlalchemy warning about retrieving an object without primary key.

**Comment by peterschutt** [comment]
This ended up a bit broader than I originally anticipated so let me know if you'd prefer that I split the issues and PRs up.

**Comment by lafrech** [comment]
Sorry for the delay.

This looks pretty good. Those warnings are annoying me too, so thanks for tackling this.

Besides, with SQLAlchemy 1.4 out, `Binary` is now removed, so the tests even fail.

> limit where info={"marshmallow: ...}" is passed to Column in tests in order to reduce incidence of depreciation alert when running tests.

IIUC, there should still be a warning but I don't see it in the test output. If there is something there to be addressed, it can be done later on anyway.

> This ended up a bit broader than I originally anticipated so let me know if you'd prefer that I split the issues and PRs up.

I'd have separated marshmallow and SQLAlchemy related changes, but I don't mind. I believe this PR is good enough for our standards and should be mergde already.

@sloria, I don't think I have the permission to merge in this repo.

@peterschutt could you mark this PR as ready or is there still things you'd like to address?

Note that `test_auto_field_works_with_synonym` fails with very latest 1.4.9 so if rebased, tests will fail, but it shouldn't prevent merging. This can be fixed in another PR.

Thanks again @peterschutt.

**Comment by sloria** [comment]
@lafrech You should have write access now.

**Comment by lafrech** [comment]
Yes, thanks.

I could make this "ready for review" and merge but let's wait for @peterschutt.

**Comment by peterschutt** [comment]
Thanks.

Given that you are happy I've removed the draft status. At the time, I just wanted to signal that I was unsure about the methodology I'd used and the scope of the changes.

I'm a little squeezed for time at the moment and it was a little while ago that I wrote this, so it all feels a bit foreign to me. If I have some time at the end of the day today I'll do another pass over the changes to refresh myself.

File: src/marshmallow_sqlalchemy/convert.py
Comment (lafrech) [reply src/marshmallow_sqlalchemy/convert.py]: s/DEPRECIATED/DEPRECATED

**Review by lafrech — CHANGES_REQUESTED** [review]
Looks good to me. Just that `DEPRECATED` typo. If you don't have time, I can fix it.

**Comment by peterschutt** [comment]
> > limit where info={"marshmallow: ...}" is passed to Column in tests in order to reduce incidence of depreciation alert when running tests.
> 
> IIUC, there should still be a warning but I don't see it in the test output. If there is something there to be addressed, it can be done later on anyway.

If I run `python -m pytest . -W error` this shows up as the first warning:

```
FAILED tests/test_conversion.py::TestModelFieldConversion::test_info_overrides - DeprecationWarning: Passing `info={"marshmallow": ...}` is deprecated. Use `SQLAlchemySchema` and `auto_field` instead.
```

File: src/marshmallow_sqlalchemy/convert.py
Comment (peterschutt) [reply src/marshmallow_sqlalchemy/convert.py]: Wow.. TIL

**Comment by lafrech** [comment]
Thanks.

As I expected, tests fail due to a change in 14.9 regarding "synonym". If you have any idea how to fix this, you're more than welcome to do it. Otherwise, maybe we could merge anyway.

Likewise, the last remaining warning could be addressed to get a warning-clear output, but it does not need to be done before merging this as this is already a great improvement.

**Comment by peterschutt** [comment]
We are using `SynonymProperty._proxied_property` to identify whether a prop is a synonym which was [renamed to `_proxied_object`](https://github.com/sqlalchemy/sqlalchemy/commit/b73fc8f874da94c9c5b2d94feb6b1b45b7f4f02b#diff-3b4c23cd96e4cca98c26a02d627673e76e04e3ca8318d568cabeae91ee145336R673) a week or two ago.

[`_proxied_object` still appears to be unique](https://github.com/sqlalchemy/sqlalchemy/search?q=_proxied_object) so I've changed it to check for both.

**Review by lafrech — APPROVED** [review]

**Comment by peterschutt** [comment]
For #361

**Comment by lafrech** [comment]
Thanks.

**Pull request merged**
